### DISCUSSION TITLE
Using LockMode.PESSIMISTIC_WRITE on owned side of OneToOne

### DISCUSSION THREAD

**Discussion by ErinCall** [mikro-orm/mikro-orm#5125] [Q&A] [answered]
Suppose I have two entities like so:

```ts
@Entity({ schema: 'clinic' })
export class Visit {
  @PrimaryKey({ type: UuidType })
  id: string = uuid();

  @OneToOne({ entity: () => VisitReason, mappedBy: 'visit' })
  visitReason?: VisitReason | null;

  /* other stuff... */
}

@Entity({ schema: 'clinic' })
export class VisitReason {
  @PrimaryKey({ type: UuidType })
  id: string = uuid();

  @OneToOne(() => Visit)
  visit!: Visit;

  /* other stuff... */
}
```

And then I do something like this:

```ts
return this.entityManager.transactional(async () => {
  const visit = await this.entityManager.findOne(
    Visit,
    { id: visitId },
    { lockMode: LockMode.PESSIMISTIC_WRITE },
  );

  /* do stuff... */
});
```

When the query runs, it will get an error (at least in postgres): `FOR UPDATE cannot be applied to the nullable side of an outer join`. Which I suppose makes sense; you can't lock a row that doesn't exist. I've found a few solutions, but they're all a bit distasteful:

* Use `select ... for update of visit`. This would be ideal but I don't see a way to do it in Mikro (tell me I'm wrong? 🥺).
* Turn off `autoJoinOneToOneOwner` globally. I'm reluctant to touch this, since it's hard to identify existing code in unrelated areas that quietly assumes that the autojoin behavior is on. There's no way to disable it for a specific query, right?
* Invert the owner/owned relationship. It's probably fine but it creates the possibility of an orphaned visit_reason
* Use optimistic locking instead of `for update`. Might be fine here but it's a lot of overhead for a case where database locking is otherwise suitable.
* Remove the `for update` entirely and accept the race condition. It's probably fine in this particular case but it might not be fine the next time. 

Do I have any other options?

**Comment by B4nan** [comment, marked as answer]
Maybe you could use `lockTableAliases` option (see the end of this section https://mikro-orm.io/docs/transactions#pessimistic-locking)? It will produce `for update of [alias list] skip locked` query.

**Reply by ErinCall** [reply]
Alas, that runs into a new problem. I didn't originally mention this in the example since I didn't realize it was relevant, but the tables are using a schema name. That oughtn't be a problem but it triggers a knex bug, knex/knex#5053 🙁

**Reply by B4nan** [reply]
Maybe try asking in that issue? It sounds like an easy fix on their end, maybe they would accept a PR.

**Reply by B4nan** [reply]
Actually, how were you trying it out? You are supposed to pass an array of aliases there, not table names, so where did the schema name get into it?

edit: ok, tried in the ORM tests and I see it does fail this way. but as expected, the fix is trivial ([this is the bit that needs change](https://github.com/knex/knex/blob/master/lib/dialects/postgres/query/pg-querycompiler.js#L222))

**Reply by ErinCall** [reply]
Ok, opened knex/knex#5791 to fix that bug. Thanks for pointing me in the right direction. 

Assuming they accept the PR, is my next step to come back here and say "hi please bump the knex dependency"?

**Reply by B4nan** [reply]
It will happen automatically via renovate, which will trigger a dev release.

**Reply by B4nan** [reply]
Btw not sure if you know https://github.com/ds300/patch-package, you could use that right now to patch knex in your project before this gets reviewed and processed.

**Reply by ErinCall** [reply]
Thanks for that, it worked like a charm 🙂

**Comment by B4nan** [comment]
Ended up including the fix from your PR in the custom dialect: https://github.com/mikro-orm/mikro-orm/commit/3fdb077893144111cf4d8b2ea5a22271d0f405df
